runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+    val allocAll = IO {
+      model.allocations.map { case (name, allocation) =>
+        val totalBytes = (allocation.shape.product * allocation.dataType.sizeInBytes).toUSize
+        val ptr = stdlib.malloc(totalBytes)
+        if (ptr == null) throw new OutOfMemoryError(s"Failed to allocate tensor '$name'")
+        allocation.initialData.foreach(data =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize),
+        )
+        name -> ptr
+      }
+    }
+    Resource.make(allocAll)(memoryMap => IO(memoryMap.values.foreach(stdlib.free))).map(_.toMap)
+  }
+
+  /** Executes a single operation by dispatching to the appropriate handler. */
+  private def executeOperation(op: Operation, memory: MemoryMap, model: ModelIR): IO[Unit] =
+    op match {
+      case op: Operation.SVMClassifier => handleSvmClassifier(op, memory, model)
+      case op: Operation.Add => handleElementWise(op, memory, model)(_ + _)
+      case op: Operation.Mul => handleElementWise(op, memory, model)(_ * _)
+      case op: Operation.Cast => handleCast(op, memory, model)
+      case other =>
+        IO.raiseError(
+          new NotImplementedError(s"Operation not implemented: ${other.getClass.getSimpleName}"),
+        )
+    }
+
+  // --- Operation Handlers ---
+
+  private def handleElementWise(op: Operation, memory: MemoryMap, model: ModelIR)(
+      f: (Float, Float) => Float,
+  ): IO[Unit] = IO {
+    val outputAllocation = model.allocations(op.outputs.head)
+    val count = outputAllocation.shape.product
+
+    val inputA = memory(op.inputs(0)).asInstanceOf[Ptr[CFloat]]
+    val inputB = memory(op.inputs(1)).asInstanceOf[Ptr[CFloat]]
+    val output = memory(op.outputs(0)).asInstanceOf[Ptr[CFloat]]
+
+    for (i <- 0 until count)
+      !(output + i) = f(!(inputA + i), !(inputB + i))
+  }
+
+  private def handleCast(op: Operation.Cast, memory: MemoryMap, model: ModelIR): IO[Unit] = IO {
+    val inputAlloc = model.allocations(op.input)
+    val outputAlloc = model.allocations(op.output)
+
+    val inputPtr = memory(op.input)
+    val outputPtr = memory(op.output)
+    val count = inputAlloc.shape.product
+
+    (inputAlloc.dataType, outputAlloc.dataType) match {
+      case (from, to) if from == to =>
+        val _ = memcpy(outputPtr, inputPtr, (count * from.sizeInBytes).toUSize)
+      case (DataType.Float64, DataType.Float32) =>
+        val in = inputPtr.asInstanceOf[Ptr[CDouble]]
+        val out = outputPtr.asInstanceOf[Ptr[CFloat]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toFloat
+      case (DataType.Float32, DataType.Float64) =>
+        val in = inputPtr.asInstanceOf[Ptr[CFloat]]
+        val out = outputPtr.asInstanceOf[Ptr[CDouble]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toDouble
+      // Add more cases for other data types
+      case (from, to) =>
+        throw new NotImplementedError(s"Cast from $from to $to is not implemented.")
```

Comment:
Instead of raising the error at runtime inside of the `IO`, we should raise this error at compile-time outside of the `IO`. So the user would know right away that their model will not compile, instead of finding out when the try to run inference.